x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/FromPartialAggregatorFunction.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.aggregation;
+
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.CompositeBlock;
+import org.elasticsearch.compute.data.ElementType;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.compute.operator.DriverContext;
+import org.elasticsearch.core.Releasables;
+
+import java.util.List;
+
+public class FromPartialAggregatorFunction implements AggregatorFunction {
+    private static final List<IntermediateStateDesc> INTERMEDIATE_STATE_DESC = List.of(
+        new IntermediateStateDesc("partial", ElementType.COMPOSITE, "partial_agg")
+    );
+
+    public static List<IntermediateStateDesc> intermediateStateDesc() {
+        return INTERMEDIATE_STATE_DESC;
+    }
+
+    private final DriverContext driverContext;
+    private final GroupingAggregatorFunction groupingAggregator;
+    private final int inputChannel;
+    private boolean receivedInput = false;
+
+    public FromPartialAggregatorFunction(DriverContext driverContext, GroupingAggregatorFunction groupingAggregator, int inputChannel) {
+        this.driverContext = driverContext;
+        this.groupingAggregator = groupingAggregator;
+        this.inputChannel = inputChannel;
+    }
+
+    @Override
+    public void addRawInput(Page page) {
+        addIntermediateInput(page);
+    }
+
+    @Override
+    public void addIntermediateInput(Page page) {
+        try (IntVector groupIds = driverContext.blockFactory().newConstantIntVector(0, page.getPositionCount())) {
+            if (page.getPositionCount() > 0) {
+                receivedInput = true;
+            }
+            final CompositeBlock inputBlock = page.getBlock(inputChannel);
+            groupingAggregator.addIntermediateInput(0, groupIds, inputBlock.asPage());
+        }
+    }
+
+    private IntVector outputPositions() {
+        return driverContext.blockFactory().newConstantIntVector(0, receivedInput ? 1 : 0);
+    }
+
+    @Override
+    public void evaluateIntermediate(Block[] blocks, int offset, DriverContext driverContext) {
+        final Block[] partialBlocks = new Block[groupingAggregator.intermediateBlockCount()];
+        boolean success = false;
+        try (IntVector selected = outputPositions()) {
+            groupingAggregator.evaluateIntermediate(partialBlocks, 0, selected);
+            blocks[offset] = new CompositeBlock(partialBlocks);
+            success = true;
+        } finally {
+            if (success == false) {
```

Comment:
I feel like if the caller is giving you an array of blocks to fill in it's probably their responsibility to close blocks on failure.

But we should close `partialBlocks` if that fails.

Reply:
Good catch. I fixed it in https://github.com/elastic/elasticsearch/pull/110206/commits/307f734986e15180f81e87a1da1def5de64799ed.  Thanks Nik!